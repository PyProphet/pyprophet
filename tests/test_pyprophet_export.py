from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
