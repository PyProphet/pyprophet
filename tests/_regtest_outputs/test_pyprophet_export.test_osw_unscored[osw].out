    Charge                  FullPeptideName  Intensity          ProteinName       RT             Sequence                           aggr_Fragment_Annotation                                     aggr_Peak_Apex                                     aggr_Peak_Area  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_iRT   assay_rt  d_score  decoy  delta_iRT  delta_rt                         filename      iRT                   id  leftWidth  m_score        mz  peak_group_rank  rightWidth               run_id  transition_group_id
0        2       ADSTGTLVITDPTR(UniMod:267)   207283.0  AQUA4SWATH_HMLangeA  2661.55       ADSTGTLVITDPTR  1359_y3_1;1344_b5_1;1360_y4_1;1345_b6_1;1361_y...  36907.0;1426.0;6131.0;2071.0;11984.0;559.0;352...  251772.0;9915.0;43365.0;15040.0;80527.0;4110.0...          117220.7482             854645.0       26.5  2595.5788   5.7301      0     1.9379   65.9712  napedro_L120420_010_SW.mzXML.gz  28.4379 -4409520928686189639  2640.5100   0.0033  728.8795                1   2705.3701 -8670811102654834151                    0
1        2        ALGYEDATQALGR(UniMod:267)   356015.0  AQUA4SWATH_HMLangeA  2915.46        ALGYEDATQALGR  1392_y3_1;1377_b4_1;1393_y4_1;1378_b5_1;1394_y...  6434.0;3256.0;10290.0;2827.0;6724.0;1335.0;205...  52418.0;29826.0;74242.0;25170.0;52600.0;15741....          184097.4993            1350180.0       36.8  2949.3521   5.6547      0    -0.9997  -33.8921  napedro_L120420_010_SW.mzXML.gz  35.8003 -7771919224870429764  2892.9299   0.0033  687.8480                1   2947.5500 -8670811102654834151                    1
2        2        APIPTALDTDSSK(UniMod:259)   446618.0  AQUA4SWATH_HMLangeA  2401.28        APIPTALDTDSSK  1423_y4_1;1410_b5_1;1404_b10_2;1418_y10_2;1424...  4851.0;1505.0;6080.0;22378.0;13531.0;1278.0;63...  35626.0;9642.0;36807.0;117103.0;78253.0;8760.0...          384029.0662            1982920.0       19.1  2341.4023   5.4280      0     1.7913   59.8777  napedro_L120420_010_SW.mzXML.gz  20.8913  -797725006165535344  2384.1599   0.0033  662.3477                1   2428.5400 -8670811102654834151                    2
3        2            DAVNDITAK(UniMod:259)   550550.0  AQUA4SWATH_HMLangeA  1440.97            DAVNDITAK  1444_y7_2;1433_b4_1;1440_y4_1;1434_b5_1;1441_y...  8915.0;2857.0;9466.0;2115.0;12085.0;48878.0;80...  48372.0;21445.0;48669.0;11461.0;64587.0;282637...          346652.2688            1718980.0       -7.8  1417.4678   5.6703      0     0.8459   23.5022  napedro_L120420_010_SW.mzXML.gz  -6.9541  8126738070040012582  1420.7200   0.0033  477.7527                1   1461.6899 -8670811102654834151                    3
4        2          DITAFDETLFR(UniMod:267)   316323.0  AQUA4SWATH_HMLangeA  4736.91          DITAFDETLFR  1450_b4_1;1461_y3_1;1467_y7_2;1462_y4_1;1451_b...  4177.0;2790.0;1081.0;6092.0;2289.0;3674.0;4943...  33058.0;18451.0;11113.0;40805.0;18855.0;30367....           89199.6087             569321.0       88.1  4711.3648   5.8401      0     0.5151   25.5452  napedro_L120420_010_SW.mzXML.gz  88.6151 -1732939685941081620  4715.8398   0.0033  669.3342                1   4770.4702 -8670811102654834151                    4
..     ...                              ...        ...                  ...      ...                  ...                                                ...                                                ...                                                ...                  ...                  ...        ...        ...      ...    ...        ...       ...                              ...      ...                  ...        ...      ...       ...              ...         ...                  ...                  ...
95       2           TFADGSLAFR(UniMod:267)   788463.0  AQUA4SWATH_HMLangeC  3138.11           TFADGSLAFR  3151_y3_1;3143_b4_1;3152_y4_1;3142_-1_1;3154_y...  8338.0;4202.0;6502.0;20900.0;32653.0;2257.0;24...  56215.0;30204.0;50445.0;198991.0;198460.0;4964...          264521.1026            2012060.0       44.2  3203.5231   5.3166      0    -1.9437  -65.4131  napedro_L120420_010_SW.mzXML.gz  42.2563 -5006126068043441226  3117.6399   0.0033  547.7789                1   3168.8401 -8670811102654834151                   65
96       2           TFADGSLAFR(UniMod:267)   137139.0  AQUA4SWATH_HMLangeC  3010.68           TFADGSLAFR                                                NaN                                                NaN                                                NaN           24208.4173             280044.0       44.2  3203.5268   2.1156      0    -5.6388 -192.8468  napedro_L120420_010_SW.mzXML.gz  38.5612 -3923479974150751688  2974.2500   0.0210  547.7789                2   3035.7000 -8670811102654834151                   65
97       2  TSDTGEEEAITDSTEANGK(UniMod:259)    38519.0  AQUA4SWATH_HMLangeC  1857.78  TSDTGEEEAITDSTEANGK  3207_y4_1;3180_b4_1;3181_b5_1;3208_y5_1;3182_b...  2968.0;1272.0;561.0;1488.0;879.0;1181.0;1820.0...  11728.0;5099.0;2298.0;5328.0;3236.0;3769.0;714...           42908.1783             195600.0        6.1  1894.8893   5.7596      0    -0.9681  -37.1093  napedro_L120420_010_SW.mzXML.gz   5.1319  5044606440762346646  1842.6700   0.0033  981.9287                1   1876.8101 -8670811102654834151                   66
98       2          ADHSGAVVLLK(UniMod:259)    37110.0  AQUA4SWATH_HMLangeD  1973.12          ADHSGAVVLLK  3229_y3_1;3238_y9_2;3216_-1_1;3222_b7_1;3233_y...  888.0;2833.0;17756.0;2603.0;1101.0;1803.0;1760...  6918.0;27993.0;195730.0;24406.0;7597.0;17242.0...           75249.4487             897889.0        9.9  2025.4137   4.1353      0    -1.4238  -52.2937  napedro_L120420_010_SW.mzXML.gz   8.4762  1099848936678289217  1943.3500   0.0062  559.3264                1   1997.9800 -8670811102654834151                   67
99       2         EDNTGVVSQTFK(UniMod:259)   550945.0  AQUA4SWATH_HMLangeD  2070.68         EDNTGVVSQTFK  3244_b3_1;3258_y3_1;3245_b4_1;3246_b5_1;3259_y...  6540.0;15973.0;6968.0;3001.0;6504.0;5393.0;447...  27695.0;66065.0;28730.0;13867.0;28784.0;23144....          395820.4672            1666220.0       10.8  2056.3210   5.8824      0     0.5052   14.3590  napedro_L120420_010_SW.mzXML.gz  11.3052 -4744925383774838539  2053.0100   0.0033  666.8297                1   2087.1499 -8670811102654834151                   68

[100 rows x 27 columns]
